File: AsianPrice.R
#Asian Option Price Function
AsianPrice = function(S0,K,vol,r,t,dt,Realizations){
  
  #Precalculate constants
  Ncols = as.numeric(ceiling(t/dt))
  Nrows = Realizations
  drift <- (r-((vol^2)/2))*dt
  partial.stochastic.part <- vol*sqrt(dt)
  DiscountFactor = rep(exp(-r*t),times=Nrows)
  
  #Simulate Stock Movements
  S <- matrix(exp(drift+partial.stochastic.part*rnorm(Ncols*Nrows)), ncol=Ncols, nrow=Nrows)
  S <- t(apply(S,1,cumprod))
  S <- S*S0
  S[,1] <- S0
  #matplot(t(S), type = "l")
  
  #Generate Average
  A.Path = (1/Ncols)*rowSums(S)
      
  #Payoffs
  Payoff.Call = pmax(A.Path - K,0)
  Payoff.Put = pmax(K-A.Path,0)
   
  #Price
  Call = mean(Payoff.Call*DiscountFactor)
  Put = mean(Payoff.Put*DiscountFactor)
  Labels = c("Call","Put")
  Premiums = numeric(2)
  Premiums[1] = Call
  Premiums[2] = Put
  Output = list(Labels, Premiums)
  return(Output)
}

File: AsianPrice.m
%Asian Option Pricing

function [AsianCall,AsianPut]=AsianPrice(S0,K,r,vol,T,dt,Nrows)

%Nrows is the number of realizations or paths, i.e. Nrows == N.

tic
    
%Pre-calculate as many parameters as possible
    Ncols = T/dt;
    drift = (r-((vol^2)/2))*dt; 
    part_stoch_part = vol*sqrt(dt); 
    Discount = exp(-r*T);

%Generate Random numbers
    random = randn(Nrows, Ncols);

%Generate Stock Price Paths
    S = ones(Nrows, Ncols);
    S = S.*exp(drift+part_stoch_part.*random);
    S = cumprod(S,2);
    S = S0*S;
    S(:,1) = S0;

clear random; %Free up a chunk of RAM

%Calculate Asian Option Prices
    %Asian Call
        %Path Averages
            A_Path = (1/Ncols)*sum(S,2);
        %Calculate Payoff
            Payoff_Call = [zeros(Nrows,1) A_Path - K];
            Payoff_Call = Discount.*max(Payoff_Call,[],2);
            AsianCall = mean(Payoff_Call);
    %Asian Put
        %Calculate Payoff
            Payoff_Put = [zeros(Nrows,1) K - A_Path];
            Payoff_Put = exp(-r*T).*max(Payoff_Put,[],2);
            AsianPut = mean(Payoff_Put);
toc
